### PR TITLE
Add-two-new-mca-variables, fix opal_help failure

### PR DESCRIPTION
**Pull request by mike-dubman**

### CODE REVIEW

**Comment by mike-dubman** [comment]
other changes are fixes for the following situation:
- All PEs should have same VA for memheap
- before fix, some sshmem allocators could end-up with having different VAs for memheap across all PEs.
- we fixed all places in sshmem allocators (hugepages, verbs) which could lead to that bug by introducing two new MCA variables. This variables can be on/off hugepages and verbs(shared_mr) services dynamically if requested VA cannot be attached and next or other allocator will be used instead.
- Please let this change as-is, we will try harder to break PRs into smaller pieces for next ones. 
  Thanks

File: oshmem/mca/sshmem/sysv/sshmem_sysv_component.c [hunk 1]
```diff
@@ -171,7 +186,14 @@ sysv_register(void)
                                            MCA_BASE_VAR_SCOPE_ALL_EQ,
                                            &mca_sshmem_sysv_component.priority);
 
-    mca_sshmem_sysv_component.use_hp = 0;
+    mca_sshmem_sysv_component.use_hp = 2;
+    mca_base_component_var_register (&mca_sshmem_sysv_component.super.base_version,
+                                           "use_hp", "Huge pages usage "
+                                           "[0 - off, 1 - on, 2 - auto] (default: 2)", MCA_BASE_VAR_TYPE_INT,
+                                           NULL, 0, MCA_BASE_VAR_FLAG_SETTABLE,
+                                           OPAL_INFO_LVL_3,
```
Comment (jsquyres) [inline oshmem/mca/sshmem/sysv/sshmem_sysv_component.c]: Is this really level 3?  By https://github.com/open-mpi/ompi/wiki/MCAParamLevels, I'd think that this was level 4 or 5.

File: oshmem/mca/sshmem/sysv/sshmem_sysv_component.c [hunk 2]
```diff
@@ -171,7 +186,14 @@ sysv_register(void)
                                            MCA_BASE_VAR_SCOPE_ALL_EQ,
                                            &mca_sshmem_sysv_component.priority);
 
-    mca_sshmem_sysv_component.use_hp = 0;
+    mca_sshmem_sysv_component.use_hp = 2;
+    mca_base_component_var_register (&mca_sshmem_sysv_component.super.base_version,
+                                           "use_hp", "Huge pages usage "
+                                           "[0 - off, 1 - on, 2 - auto] (default: 2)", MCA_BASE_VAR_TYPE_INT,
```
Comment (jsquyres) [inline oshmem/mca/sshmem/sysv/sshmem_sysv_component.c]: Suggestion: make "auto" be a value of -1; there's several other MCA params that follow this convention (i.e., that -1 == automatic).

File: oshmem/mca/sshmem/sysv/sshmem_sysv_component.c [hunk 3]
```diff
@@ -111,15 +111,30 @@ sysv_runtime_query(mca_base_module_t **module,
...
+            }
+            mca_sshmem_sysv_component.use_hp = 0;
+        }
     }
-    else if ((void *)-1 == (addr = shmat(shmid, NULL, 0))) {
-        shmctl(shmid, IPC_RMID, NULL );
+#else
+    if (mca_sshmem_sysv_component.use_hp == 1) {
         mca_sshmem_sysv_component.use_hp = 0;
+        goto out;
```
Comment (jsquyres) [inline oshmem/mca/sshmem/sysv/sshmem_sysv_component.c:141]: Shouldn't it be an error if the user asked for this functionality (i.e., use_hp==1) but you can't deliver it?  It should only be silently ignored if use_hp indicates "automatic" use.

Right?

File: oshmem/mca/sshmem/sysv/sshmem_sysv.h [hunk 4]
```diff
@@ -36,6 +38,36 @@ typedef struct mca_sshmem_sysv_module_t {
 } mca_sshmem_sysv_module_t;
 extern mca_sshmem_sysv_module_t mca_sshmem_sysv_module;
 
+/*
+ * Get current huge page size
+ */
+static size_t oshmem_gethugepagesize(void)
```
Comment (jsquyres) [inline oshmem/mca/sshmem/sysv/sshmem_sysv.h]: Why is this function defined in a header file?  I don't see it listed as inline, either, which makes it extra weird.

Is this supposed to be inline -- i.e., used in critical performance paths?

If not, it should be in a .c file.

File: oshmem/mca/sshmem/verbs/sshmem_verbs_component.c [hunk 5]
```diff
@@ -285,6 +291,15 @@ verbs_register(void)
                                          MCA_BASE_VAR_SYN_FLAG_DEPRECATED);
     }
 
+    mca_sshmem_verbs_component.has_shared_mr = 2;
+    index = mca_base_component_var_register (&mca_sshmem_verbs_component.super.base_version,
+                                           "shared_mr", "Shared memory region usage "
+                                           "[0 - off, 1 - on, 2 - auto] (default: 2)", MCA_BASE_VAR_TYPE_INT,
```
Comment (jsquyres) [inline oshmem/mca/sshmem/verbs/sshmem_verbs_component.c]: Same suggestion as above: use -1 to indicate "automatic" behavior.

File: oshmem/mca/sshmem/verbs/sshmem_verbs_component.c [hunk 6]
```diff
@@ -285,6 +291,15 @@ verbs_register(void)
...
+                                           NULL, 0, MCA_BASE_VAR_FLAG_SETTABLE,
+                                           OPAL_INFO_LVL_3,
```
Comment (jsquyres) [inline oshmem/mca/sshmem/verbs/sshmem_verbs_component.c]: Same suggestion as above: sounds like this should be a level 4 or 5.

File: oshmem/mca/sshmem/sysv/sshmem_sysv_component.c (hunk 1 above)
Comment (mike-dubman) [inline oshmem/mca/sshmem/sysv/sshmem_sysv_component.c]: fixed to level4

File: oshmem/mca/sshmem/sysv/sshmem_sysv_component.c (hunk 2 above)
Comment (mike-dubman) [inline oshmem/mca/sshmem/sysv/sshmem_sysv_component.c]: using -1 now as suggested

File: oshmem/mca/sshmem/sysv/sshmem_sysv_component.c (hunk 3 above)
Comment (mike-dubman) [inline oshmem/mca/sshmem/sysv/sshmem_sysv_component.c:141]: fixed, thanks

File: oshmem/mca/sshmem/sysv/sshmem_sysv.h (hunk 4 above)
Comment (mike-dubman) [inline oshmem/mca/sshmem/sysv/sshmem_sysv.h]: fixed, thanks

File: oshmem/mca/sshmem/verbs/sshmem_verbs_component.c (hunk 5 above)
Comment (mike-dubman) [inline oshmem/mca/sshmem/verbs/sshmem_verbs_component.c]: using -1, thanks

File: oshmem/mca/sshmem/verbs/sshmem_verbs_component.c (hunk 6 above)
Comment (mike-dubman) [inline oshmem/mca/sshmem/verbs/sshmem_verbs_component.c]: using 4, thanks

**Comment by mike-dubman** [comment]
- Will open separate issue to s/#ifdef MACRO/#if MACRO/g for @nkogteva 
- Will re-integrate into "master" once no more comments observed.

**Comment by rhc54** [comment]
What is the status of this PR? I see last comments indicating further work required, and that these changes may not yet be in the trunk. True?

**Comment by mike-dubman** [comment]
@rhc54 - all comments were addressed, PR is ready to go.

**Comment by rhc54** [comment]
Where is the master hash?

**Comment by mike-dubman** [comment]
open-mpi/ompi@5ce2f1006762ef908b2960df5b93dc7532f8e448

**Comment by mike-dubman** [comment]
attempt 2:
open-mpi/ompi@5ce2f1006762ef908b2960df5b93dc7532f8e448

**Comment by rhc54** [comment]
There are conflicts in this patch - please resolve them.

**Comment by mike-dubman** [comment]
@rhc54 - done. Thanks.

**Pull request merged**